### PR TITLE
NET DATUMS!!!!1

### PR DESCRIPTION
**Pull request by PaulRitter**
So yeah this is me reworking ALL nets to be datums, so far i wrote the basetypes of a net and a node, but i imagine those will change after i rework the powernet to be a childtype and realize some of my ideas are limiting.

This would also make it WAY more easier to make ANY form of net to track fungi or kudzu or Powercreeper #23458 (shameless selfplug).

This PR was brought to you by **me looking at gunky cable code and wanting to kill myself**:tm:

- [X] add base net, node 
- rework anything fucking with power
  - [x] cables
  - [x] power machines
  - [x] batteries
  - [x] power monitor
  - [ ] powersinks
  - [ ] apc
  - [ ] smes and terminals in general
  - [ ] help me please
- [x] rework the SS
- [x] add power net
- [x] /proc/electrocute_mob()
- [x] replace some avail()'s with excess() cause my brain goofed
- [ ] create wiki tutorial on creating nets
- [ ] review my own code to see if i fucked up somewhere

[wip]
:cl:
 * experiment: All nets are now entirely made out of datums, what could go wrong

### CODE REVIEW

File: code/datums/net/base/net.dm [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    if(!istype(src, node.netType))
+        return 0
+
+    node.net = src
+    nodes |= node
+    return 1
+
+//merges with another net
+/datum/net/proc/absorb_net(var/datum/net/other_net)
```
Comment (ShiftyRail) [inline code/datums/net/base/net.dm:34]: Could probably also qdel the other net.

File: code/datums/net/base/net_atom.dm [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/atom
+	var/list/net_nodes = list() //list of nodes
+
+/atom/New(loc)
+	. = ..(loc)
```
Comment (ShiftyRail) [inline code/datums/net/base/net_atom.dm]: This does nothing, I think

**Comment by DeityLink** [comment]
I thought you were talking about bug nets, got me confused for a moment.

File: code/datums/net/base/net_atom.dm (hunk 2 above)
Comment (PaulRitter) [reply code/datums/net/base/net_atom.dm]: needed since there are multiple /atom/New()'s in the code, apparently they dont get called if you dont supercall

Comment (DamianX) [reply code/datums/net/base/net_atom.dm]: No, he's right. This  does nothing.

Comment (PaulRitter) [reply code/datums/net/base/net_atom.dm]: it does, i pinged you in discord

File: code/datums/net/base/net.dm [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+//used in merge_nets proc to determine which network needs to be absorbed (the smaller one)
+//this returns a number
+/datum/net/proc/get_size()
+    return nodes.len
+
+//returns one net
+/proc/merge_nets(var/datum/net/n1, var/datum/net/n2)
+    if(!n1)
+        return n2 ? n2 : null
```
Comment (ShiftyRail) [inline code/datums/net/base/net.dm]: This is identical to `return n2`

File: code/datums/net/power/net.dm [hunk 4]
```diff
@@ -0,0 +1,18 @@
+/datum/net/power
+    note_types = list(/datum/net_node/power)
+    var/load = 0				// the current load on the powernet, increased by each machine at processing
+    var/newavail = 0			// what available power was gathered last tick, then becomes...
+    var/avail = 0				// ...the current available power in the powernet
+    var/viewload = 0			// the load as it appears on the power console (gradually updated)
+    var/netexcess = 0			// excess power on the powernet (typically avail-load)
+
+//merges with another net
+/datum/net/power/absorb_net(var/datum/net/power/other_net)
```
Comment (ShiftyRail) [inline code/datums/net/power/net.dm:16]: Should probably return something if it worked.

**Comment by Kurfursten** [comment]
You may request to have this reopened

**Pull request closed without merging**